filters/openpolicyagent/openpolicyagent.go

```diff
@@ -470,6 +485,42 @@ func (opa *OpenPolicyAgentInstance) MetricsKey(key string) string {
 	return key + "." + opa.bundleName
 }
 
+type bufferedReadCloser struct {
+	reader *bufio.Reader
+	closer io.ReadCloser
+}
+
+func (r *bufferedReadCloser) Read(p []byte) (n int, err error) {
+	return r.reader.Read(p)
+}
+
+func (r *bufferedReadCloser) Close() error {
+	return r.closer.Close()
+}
+
+func (opa *OpenPolicyAgentInstance) ExtractHttpBodyOptionally(req *http.Request) (io.ReadCloser, []byte, error) {
+	body := req.Body
+	var rawBody []byte
+	if body != nil && !opa.EnvoyPluginConfig().SkipRequestBodyParse && opa.maxBodyBytes > 0 {
+		bufferedReader := bufio.NewReaderSize(body, opa.maxBodyBytes)
+		wrapper := &bufferedReadCloser{
+			closer: req.Body,
+			reader: bufferedReader,
+		}
+
+		var err error
+		rawBody, err = bufferedReader.Peek(opa.maxBodyBytes)
```

Comment:
That's likely blocking call to read(2) and can be used as DoS vector.
Can you please make sure it's not blocking, but streaming in case you need to work on the request body?

Reply:
I would suggest to create another filter for the body inspection.
Then for the implementation it would be a streaming filter that works on the body and as soon as it is clear that the body should be denied stop by returning an error, similar to blockContent filter.
For details how to do streaming filters: https://github.com/zalando/skipper/pull/2428 , I also had some more diret pointers in chat.

Reply:
@szuecs I have some questions
With streaming do you mean that
1. We read the body partially like the blockContent filter does?
2. With streaming is the expectation to still hit the backend before the request filter finishes the processing?

Reply:
Thanks for the pointers @szuecs, seems doable after reading the linked PR, just a few clarifying questions (and slightly separate topics)

(a) Separate Filter
Is the motivation to not touch the body unless needed or is there some other thing that is missing? 
It should be possible to be a bit smarter in the filter and only parse the body only if it is really needed by the policy (needs some exploration with partial evaluation) and it fits from a size perspective. From a UX perspective I think ending up with different filters for each use case could be confusing but it is also a bit hard to predict which other cases we'll see.  This would also be a good motivator to include a benchmark...

(b) Streaming body
Given that the mentioned PR is not yet merged, this essentially would result in a special purpose wrapper, do you think this is ok and we try to unify later or should we wait for the PR to be merged?
Secondly, the policy can influence the returned body / headers. If the deny happens during the body processing, I am not yet sure where we would implement that. I assume that the error returned during `Read()` is somehow propagated to the `Response()` method of the filter or is there a better way for this?

Reply:
(a) if it's possible to know at parse time that you need to touch the body, then it would be great to only touch the body in case you need.
(b) yes it would already send the request headers to the backend and start streaming, but your filter will be able to stop the body streaming, because it gets the bytes before these will be written into the wire. You can pass errors to the Response of the filter via filtercontext's stateBag (map[string]interface{}) .

Reply:
Thanks, then I'll proceed with making the changes:
(a) will try to get this info, if that is not feasible or slow, we can default to the separate filter
(b) will switch to a streaming processing then with the information you provided...

Reply:
When starting to implement (b), I encountered an issue that I missed: The decision can influence the headers that are sent to the downstream service and we have an internal use case that will rely on that. 
I assume that the actual `Read` for passing the request is actually done by the `http.Client` that makes the outgoing request. This client actually copies the headers before starting to read the body which means that they can not longer be influenced. 
I am now a bit unsure how to proceed. (a) seems possible after initial testing and the read buffer can be optimised so that parsing the body does not always consume the max body size but this would still block in the `Request()` method of the filter. I will reach out via internal chat to discuss options...